ably/proto/stat.go

```diff
@@ -69,3 +21,125 @@ var (
+	Opened  float64 `json:"opened" codec:"opened"`
+	Failed  float64 `json:"failed" codec:"failed"`
+	Refused float64 `json:"refused" codec:"refused"`
+}
+
+type ConnectionTypes struct {
+	All   ResourceCount `json:"all" codec:"all"`
+	Plain ResourceCount `json:"plain" codec:"plain"`
+	TLS   ResourceCount `json:"tls" codec:"tls"`
+}
+
+type MessageCount struct {
+	Count   float64 `json:"count" codec:"count"`
+	Data    float64 `json:"data" codec:"data"`
+	Failed  float64 `json:"failed" codec:"failed"`
+	Refused float64 `json:"refused" codec:"refused"`
+}
+
+type MessageTypes struct {
+	All      MessageCount `json:"all" codec:"all"`
+	Plain    MessageCount `json:"plain" codec:"plain"`
+	Presence MessageCount `json:"presence" codec:"presence"`
+}
+
+type MessageTraffic struct {
+	All           MessageTypes `json:"all" codec:"all"`
+	RealTime      MessageTypes `json:"realtime" codec:"realtime"`
+	Rest          MessageTypes `json:"rest" codec:"rest"`
+	Webhook       MessageTypes `json:"webhook" codec:"webhook"`
+	Push          MessageTypes `json:"push" codec:"push"`
+	ExternalQueue MessageTypes `json:"externalQueue" codec:"externalQueue"`
+	SharedQueue   MessageTypes `json:"sharedQueue" codec:"sharedQueue"`
+	HTTPEvent     MessageTypes `json:"httpEvent" codec:"httpEvent"`
+}
+
+type RequestCount struct {
+	Failed    float64 `json:"failed" codec:"failed"`
+	Refused   float64 `json:"refused" codec:"refused"`
+	Succeeded float64 `json:"succeeded" codec:"succeeded"`
+}
+
+type PushTransportTypeCounter struct {
+	Total float64 `json:"total" codec:"total"`
+	GCM   float64 `json:"gcm" codec:"gcm"`
+	FCM   float64 `json:"fcm" codec:"fcm"`
+	APNS  float64 `json:"apns" codec:"apns"`
+	Web   float64 `json:"web" codec:"web"`
+}
+
+type PushNotificationFailures struct {
+	Retriable PushTransportTypeCounter `json:"retriable" codec:"retriable"`
+	Final     PushTransportTypeCounter `json:"final" codec:"final"`
+}
+
+type PushNotifications struct {
+	Invalid    float64                  `json:"invalid" codec:"invalid"`
+	Attempted  PushTransportTypeCounter `json:"attempted" codec:"attempted"`
+	Successful PushTransportTypeCounter `json:"successful" codec:"successful"`
+	Failed     PushNotificationFailures `json:"failed" codec:"failed"`
+}
+
+type PushStats struct {
+	Messages        float64           `json:"messages" codec:"messages"`
+	Notifications   PushNotifications `json:"notifications" codec:"notifications"`
+	DirectPublishes float64           `json:"directPublishes" codec:"directPublishes"`
+}
+
+type MessageDirections struct {
+	All      MessageTypes   `json:"all" codec:"all"`
+	Inbound  MessageTraffic `json:"inbound" codec:"inbound"`
+	Outbound MessageTraffic `json:"outbound" codec:"outbound"`
+}
+
+type XchgMessages struct {
+	All          MessageTypes `json:"all" codec:"all"`
+	ProducerPaid MessageTypes `json:"producerPaid" codec:"producerPaid"`
+	ConsumerPaid MessageTypes `json:"consumerPaid" codec:"consumerPaid"`
+}
+
+type BaseStats struct {
```

Comment:
This is superflous here, unless you can make `Stats` extend `BaseStats` and remove the common fields.

Reply:
my bad, will remove that .